core/emulate.c

```diff
@@ -0,0 +1,978 @@
+    /* 0x18 - 0x1F */
+    F6_ALU(em_sbb, 0), X2(N),
+    /* 0x20 - 0x27 */
+    F6_ALU(em_and, 0), X2(N),
+    /* 0x28 - 0x2F */
+    F6_ALU(em_sub, 0), X2(N),
+    /* 0x30 - 0x37 */
+    F6_ALU(em_xor, 0), X2(N),
+    /* 0x38 - 0x3F */
+    F6_ALU(em_cmp, 0), X2(N),
+    /* 0x40 - 0x47 */
+    X8(F(em_inc, op_modrm_reg, op_none, op_none, 0)),
+    /* 0x48 - 0x4F */
+    X8(F(em_dec, op_modrm_reg, op_none, op_none, 0)),
+    /* 0x50 - 0x7F */
+    X16(N), X16(N), X16(N),
+    /* 0x80 - 0x8F */
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, 0),
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group1, op_modrm_rm, op_simm8, op_none, 0),
+    X4(N),
+    I2_BV(em_mov, op_modrm_rm, op_modrm_reg, op_none, INSN_MODRM | INSN_MOV),
+    I2_BV(em_mov, op_modrm_reg, op_modrm_rm, op_none, INSN_MODRM | INSN_MOV),
+    X4(N),
+    /* 0x90 - 0x9F */
+    X16(N),
+    /* 0xA0 - 0xAF */
+    I2_BV(em_mov, op_acc, op_moffs, op_none, INSN_MOV),
+    I2_BV(em_mov, op_moffs, op_acc, op_none, INSN_MOV),
+    I2_BV(em_mov, op_di, op_si, op_none, INSN_MOV | INSN_REP | INSN_TWOMEM), /* movs{b,w,d,q} */
+    X2(N),
+    X2(N),
+    I2_BV(em_mov, op_di, op_acc, op_none, INSN_MOV | INSN_REP), /* stos{b,w,d,q} */
+    I2_BV(em_mov, op_acc, op_si, op_none, INSN_MOV | INSN_REP), /* lods{b,w,d,q} */
+    X2(N),
+    /* 0xB0 - 0xBF */
+    X16(N),
+    /* 0xC0 - 0xCF */
+    X4(N),
+    X2(N),
+    G(opcode_group11, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group11, op_modrm_rm, op_imm, op_none, 0),
+    X8(N),
+    /* 0xD0 - 0xEF */
+    X16(N), X16(N),
+    /* 0xF0 - 0xFF */
+    X4(N),
+    X2(N),
+    G(opcode_group3, op_none, op_none, op_none, INSN_BYTEOP),
+    G(opcode_group3, op_none, op_none, op_none, 0),
+    X8(N),
+};
+
+static const struct em_opcode_t opcode_table_0F[256] = {
+    /* 0x00 - 0xAF */
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N),
+    /* 0xB0 - 0xBF */
+    X4(N),
+    X2(N),
+    I2_BV(em_movzx, op_modrm_reg, op_modrm_rm, op_none, INSN_MODRM | INSN_MOV),
+    X8(N),
+    /* 0xC0 - 0xFF */
+    X16(N), X16(N), X16(N), X16(N),
+};
+
+static const struct em_opcode_t opcode_table_0F38[256] = {
+    /* 0x00 - 0xFF */
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+};
+
+static const struct em_opcode_t opcode_table_0F3A[256] = {
+    /* 0x00 - 0xFF */
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+};
+
+/* Emulate accesses to guest memory */
+static bool is_translation_required(struct em_context_t *ctxt)
+{
+    const struct em_opcode_t *opcode = &ctxt->opcode;
+
+    if (opcode->flags & INSN_TWOMEM) {
+        return true;
+    }
+    if (opcode->flags & INSN_STRING && ctxt->rep) {
```

Comment:
Why do `REP` instructions require translation? E.g. if the destination of `REP STOS` is an MMIO region, we'll get as many EPT violations as the repetition count (*CX), right? Each time we should be able to just use the VMCS GPA instead of translating ES:*DI to GPA, no?

Reply:
> Why do `REP` instructions require translation?

Because with each iteration ES:*DI changes (`+= (!DF ? +1 : -1) * ctxt->address_size`), we will have access memory at different GPA's, but the optimization only works when the instruction accesses a single GPA. Additionally these GPAs can't be trivially computed by just increasing/decreasing the first GPA, since we might hit different pages.

Reply:
I see. I was assuming we would resume guest execution at each iteration, which is what the old instruction emulator used to do. But it seems we are now bouncing back and forth between QEMU and the new emulator without entering guest mode at all, until we finish emulating the entire `REP` instruction. The two approaches differ in how many times we ask the hardware to execute the instruction for us (previously *CX times, now only once). Both work, but the new approach, as I now realize, is exactly the design you illustrated (i.e. the little arrows) ;-)